# **Taller #1 Modelos Dimensionales**
## Andres Felipe Hernandez Rojas

In [108]:
# Importar las librerias necesarias

import pandas as pd
import requests
import sqlite3

In [109]:
# ===================================================================
# PASO 1: CONFIGURACIÓN E INSTALACIÓN DE LIBRERÍAS
# ===================================================================

import pandas as pd
import requests
import sqlite3

print("✅ Librerías importadas.")

# ===================================================================
# PASO 2: EXTRACCIÓN (EXTRACT) DE LOS DATOS
# ===================================================================

# La plataforma datos.gov.co usa la API de Socrata. Podemos usarla para
# descargar los datos directamente, lo que es más eficiente que bajar un CSV.
# Aumentamos el límite para traer más filas (ajusta si es necesario).
api_url = "https://www.datos.gov.co/resource/nudc-7mev.json?$limit=50000"

print(f"📥 Extrayendo datos desde: {api_url}")

try:
    response = requests.get(api_url)
    response.raise_for_status()  # Lanza un error si la petición falla (ej: 404)
    data = response.json()
    df_raw = pd.DataFrame(data)
    print(f"✅ ¡Extracción exitosa! Se cargaron {len(df_raw)} filas.")
    display(df_raw.head())

except requests.exceptions.RequestException as e:
    print(f"❌ Error al extraer los datos: {e}")
    df_raw = pd.DataFrame() # Creamos un dataframe vacío para evitar errores posteriores

except Exception as e:
    print(f"❌ Ocurrió un error inesperado: {e}")
    df_raw = pd.DataFrame()

✅ Librerías importadas.
📥 Extrayendo datos desde: https://www.datos.gov.co/resource/nudc-7mev.json?$limit=50000
✅ ¡Extracción exitosa! Se cargaron 14585 filas.


a_o c_digo_municipio   municipio c_digo_departamento departamento  \
0  2023            05004    Abriaquí                  05    Antioquia   
1  2023            95025  El Retorno                  95     Guaviare   
2  2023            95200  Miraflores                  95     Guaviare   
3  2023            97001        Mitú                  97       Vaupés   
4  2023            97161      Caruru                  97       Vaupés   

  c_digo_etc              etc poblaci_n_5_16 tasa_matriculaci_n_5_16  \
0       3758  Antioquia (ETC)            503                   62.62   
1       3830   Guaviare (ETC)           4438                   53.27   
2       3830   Guaviare (ETC)           2014                   32.52   
3       3831     Vaupés (ETC)          10986                   59.57   
4       3831     Vaupés (ETC)           1228                    51.3   

  cobertura_neta  ... reprobaci_n_primaria reprobaci_n_secundaria  \
0          62.62  ...                 1.96                  16.51   
1          53.27  ...                 7.11                   9.39   
2          32.52  ...                 6.93                  14.13   
3          59.57  ...                 4.04                   8.33   
4           51.3  ...                 7.32                  15.28   

  reprobaci_n_media repitencia repitencia_transici_n repitencia_primaria  \
0              2.04       9.52                     0               10.46   
1              1.75       9.34                  6.95               11.84   
2              7.81       8.65                  6.67                9.04   
3               4.6      16.18                  7.75               21.04   
4              7.27       9.24                  2.86                7.62   

  repitencia_secundaria repitencia_media tama_o_promedio_de_grupo  \
0                 13.76             2.04                      NaN   
1                  8.48             3.16                      NaN   
2                 10.25             1.54                      NaN   
3                 13.84             7.18                      NaN   
4                 14.85             3.64                      NaN   

  sedes_conectadas_a_internet  
0                         NaN  
1                         NaN  
2                         NaN  
3                         NaN  
4                         NaN  

[5 rows x 41 columns]

In [110]:
df_raw

a_o c_digo_municipio    municipio c_digo_departamento departamento  \
0      2023            05004     Abriaquí                  05    Antioquia   
1      2023            95025   El Retorno                  95     Guaviare   
2      2023            95200   Miraflores                  95     Guaviare   
3      2023            97001         Mitú                  97       Vaupés   
4      2023            97161       Caruru                  97       Vaupés   
...     ...              ...          ...                 ...          ...   
14580  2011             5036  Angelópolis                   5    Antioquia   
14581  2011             5034        Andes                   5    Antioquia   
14582  2011             5031       Amalfi                   5    Antioquia   
14583  2011             5030        Amagá                   5    Antioquia   
14584  2011             5021   Alejandría                   5    Antioquia   

      c_digo_etc              etc poblaci_n_5_16 tasa_matriculaci_n_5_16  \
0           3758  Antioquia (ETC)            503                   62.62   
1           3830   Guaviare (ETC)           4438                   53.27   
2           3830   Guaviare (ETC)           2014                   32.52   
3           3831     Vaupés (ETC)          10986                   59.57   
4           3831     Vaupés (ETC)           1228                    51.3   
...          ...              ...            ...                     ...   
14580       3758  Antioquia (ETC)           1707                   78.85   
14581       3758  Antioquia (ETC)          10244                   84.45   
14582       3758  Antioquia (ETC)           5552                   97.71   
14583       3758  Antioquia (ETC)           6631                   78.65   
14584       3758  Antioquia (ETC)            702                  118.52   

      cobertura_neta  ... reprobaci_n_primaria reprobaci_n_secundaria  \
0              62.62  ...                 1.96                  16.51   
1              53.27  ...                 7.11                   9.39   
2              32.52  ...                 6.93                  14.13   
3              59.57  ...                 4.04                   8.33   
4               51.3  ...                 7.32                  15.28   
...              ...  ...                  ...                    ...   
14580           78.9  ...                 3.61                    9.5   
14581           84.5  ...                 0.58                   0.04   
14582           97.7  ...                    0                      0   
14583           78.7  ...                 6.73                  14.46   
14584          118.5  ...                 2.16                   4.39   

      reprobaci_n_media repitencia repitencia_transici_n repitencia_primaria  \
0                  2.04       9.52                     0               10.46   
1                  1.75       9.34                  6.95               11.84   
2                  7.81       8.65                  6.67                9.04   
3                   4.6      16.18                  7.75               21.04   
4                  7.27       9.24                  2.86                7.62   
...                 ...        ...                   ...                 ...   
14580              7.32       0.71                     0                 0.7   
14581              2.69       5.41                  0.73                5.53   
14582                 0        NaN                  0.83                 NaN   
14583              7.45       0.42                     0                0.24   
14584              8.04        0.6                     0                0.96   

      repitencia_secundaria repitencia_media tama_o_promedio_de_grupo  \
0                     13.76             2.04                      NaN   
1                      8.48             3.16                      NaN   
2                     10.25             1.54                      NaN   
3                     13.84 

La base de datos contiene 14.585 registros y 41 variables relacionadas con estadísticas educativas en Colombia para los niveles de preescolar, básica y media, desagregadas por municipio y año. Incluye información clave sobre cobertura educativa (como tasa de matriculación, cobertura neta y bruta), eficiencia del sistema (tasas de aprobación, reprobación, repitencia y deserción) y aspectos de infraestructura como el tamaño promedio de grupo y sedes con conexión a internet. Las variables geográficas (MUNICIPIO, DEPARTAMENTO, ETC) permiten realizar análisis por regiones. Se identifican algunos valores faltantes importantes, especialmente en variables de infraestructura, que deberán ser tratados en análisis posteriores. 

In [111]:
# ===================================================================
# PASO 3: EXTRACCIÓN BASE DIVIPOLA
# ===================================================================

print("📥 Extrayendo datos de DIVIPOLA desde su API...")

api_url_divipola = "https://www.datos.gov.co/resource/gdxc-w37w.json?$limit=2000"

try:
    response = requests.get(api_url_divipola)
    response.raise_for_status()
    data = response.json()
    df_divipola = pd.DataFrame(data)
    print(f"✅ ¡Base DIVIPOLA cargada! {len(df_divipola)} filas.")
except requests.exceptions.RequestException as e:
    print(f"❌ Error al extraer DIVIPOLA: {e}")
    df_divipola = pd.DataFrame()
except Exception as e:
    print(f"❌ Error inesperado: {e}")
    df_divipola = pd.DataFrame()


📥 Extrayendo datos de DIVIPOLA desde su API...
✅ ¡Base DIVIPOLA cargada! 1122 filas.


In [112]:
df_divipola

cod_dpto       dpto cod_mpio        nom_mpio          tipo_municipio  \
0          05  ANTIOQUIA    05001        MEDELLÍN               Municipio   
1          05  ANTIOQUIA    05002       ABEJORRAL               Municipio   
2          05  ANTIOQUIA    05004        ABRIAQUÍ               Municipio   
3          05  ANTIOQUIA    05021      ALEJANDRÍA               Municipio   
4          05  ANTIOQUIA    05030           AMAGÁ               Municipio   
...       ...        ...      ...             ...                     ...   
1117       97     VAUPÉS    97889        YAVARATÉ  Área no municipalizada   
1118       99    VICHADA    99001  PUERTO CARREÑO               Municipio   
1119       99    VICHADA    99524    LA PRIMAVERA               Municipio   
1120       99    VICHADA    99624   SANTA ROSALÍA               Municipio   
1121       99    VICHADA    99773        CUMARIBO               Municipio   

        longitud   latitud  
0     -75,581775  6,246631  
1     -75,428739  5,789315  
2     -76,064304  6,632282  
3     -75,141346  6,376061  
4     -75,702188  6,038708  
...          ...       ...  
1117  -69,203337  0,609142  
1118  -67,487095  6,186636  
1119  -70,410515  5,486309  
1120  -70,859499  5,136393  
1121  -69,795533  4,446352  

[1122 rows x 7 columns]

La base DIVIPOLA (División Político-Administrativa de Colombia) es el sistema oficial del DANE que identifica y clasifica todos los niveles geográficos del país. Contiene códigos y nombres para departamentos, municipios, corregimientos, centros poblados y cabeceras municipales, permitiendo una estandarización geográfica en bases de datos públicas. Cada municipio está representado por un código único de seis dígitos, donde los dos primeros identifican el departamento y los cuatro restantes el municipio.

Utilizaremos esta base de datos para hacer una de las dimensiones de nuestro modelo 

In [146]:
datos_2005_2019 = pd.read_excel("C:/Users/FEPROCOL/Desktop/Diplomado/Streamlite/diplomado-gestion-datos-dashboards-streamlit/Datos/Info_2005_2019.xlsx")
datos_2020_2035 = pd.read_excel("C:/Users/FEPROCOL/Desktop/Diplomado/Streamlite/diplomado-gestion-datos-dashboards-streamlit/Datos/Info_2020_2035.xlsx")

## Limpieza de datos

In [113]:
df_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14585 entries, 0 to 14584
Data columns (total 41 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   a_o                          14585 non-null  object
 1   c_digo_municipio             14585 non-null  object
 2   municipio                    14585 non-null  object
 3   c_digo_departamento          14585 non-null  object
 4   departamento                 14585 non-null  object
 5   c_digo_etc                   14585 non-null  object
 6   etc                          14585 non-null  object
 7   poblaci_n_5_16               14579 non-null  object
 8   tasa_matriculaci_n_5_16      14470 non-null  object
 9   cobertura_neta               14474 non-null  object
 10  cobertura_neta_transici_n    14533 non-null  object
 11  cobertura_neta_primaria      14494 non-null  object
 12  cobertura_neta_secundaria    14491 non-null  object
 13  cobertura_neta_media         14

In [114]:
pd.DataFrame(df_raw.columns.tolist(), columns=["Columnas originales"])


Columnas originales
0                           a_o
1              c_digo_municipio
2                     municipio
3           c_digo_departamento
4                  departamento
5                    c_digo_etc
6                           etc
7                poblaci_n_5_16
8       tasa_matriculaci_n_5_16
9                cobertura_neta
10    cobertura_neta_transici_n
11      cobertura_neta_primaria
12    cobertura_neta_secundaria
13         cobertura_neta_media
14              cobertura_bruta
15   cobertura_bruta_transici_n
16     cobertura_bruta_primaria
17   cobertura_bruta_secundaria
18        cobertura_bruta_media
19                    deserci_n
20         deserci_n_transici_n
21           deserci_n_primaria
22         deserci_n_secundaria
23              deserci_n_media
24                   aprobaci_n
25        aprobaci_n_transici_n
26          aprobaci_n_primaria
27        aprobaci_n_secundaria
28             aprobaci_n_media
29                  reprobaci_n
30       reprobaci_n_transici_n
31         reprobaci_n_primaria
32       reprobaci_n_secundaria
33            reprobaci_n_media
34                   repitencia
35        repitencia_transici_n
36          repitencia_primaria
37        repitencia_secundaria
38             repitencia_media
39     tama_o_promedio_de_grupo
40  sedes_conectadas_a_internet

Como NACIONAL no es un departamento lo podemos borrar

In [116]:
df_raw[df_raw['departamento']!= 'NACIONAL']

a_o c_digo_municipio    municipio c_digo_departamento departamento  \
0      2023            05004     Abriaquí                  05    Antioquia   
1      2023            95025   El Retorno                  95     Guaviare   
2      2023            95200   Miraflores                  95     Guaviare   
3      2023            97001         Mitú                  97       Vaupés   
4      2023            97161       Caruru                  97       Vaupés   
...     ...              ...          ...                 ...          ...   
14580  2011             5036  Angelópolis                   5    Antioquia   
14581  2011             5034        Andes                   5    Antioquia   
14582  2011             5031       Amalfi                   5    Antioquia   
14583  2011             5030        Amagá                   5    Antioquia   
14584  2011             5021   Alejandría                   5    Antioquia   

      c_digo_etc              etc poblaci_n_5_16 tasa_matriculaci_n_5_16  \
0           3758  Antioquia (ETC)            503                   62.62   
1           3830   Guaviare (ETC)           4438                   53.27   
2           3830   Guaviare (ETC)           2014                   32.52   
3           3831     Vaupés (ETC)          10986                   59.57   
4           3831     Vaupés (ETC)           1228                    51.3   
...          ...              ...            ...                     ...   
14580       3758  Antioquia (ETC)           1707                   78.85   
14581       3758  Antioquia (ETC)          10244                   84.45   
14582       3758  Antioquia (ETC)           5552                   97.71   
14583       3758  Antioquia (ETC)           6631                   78.65   
14584       3758  Antioquia (ETC)            702                  118.52   

      cobertura_neta  ... reprobaci_n_primaria reprobaci_n_secundaria  \
0              62.62  ...                 1.96                  16.51   
1              53.27  ...                 7.11                   9.39   
2              32.52  ...                 6.93                  14.13   
3              59.57  ...                 4.04                   8.33   
4               51.3  ...                 7.32                  15.28   
...              ...  ...                  ...                    ...   
14580           78.9  ...                 3.61                    9.5   
14581           84.5  ...                 0.58                   0.04   
14582           97.7  ...                    0                      0   
14583           78.7  ...                 6.73                  14.46   
14584          118.5  ...                 2.16                   4.39   

      reprobaci_n_media repitencia repitencia_transici_n repitencia_primaria  \
0                  2.04       9.52                     0               10.46   
1                  1.75       9.34                  6.95               11.84   
2                  7.81       8.65                  6.67                9.04   
3                   4.6      16.18                  7.75               21.04   
4                  7.27       9.24                  2.86                7.62   
...                 ...        ...                   ...                 ...   
14580              7.32       0.71                     0                 0.7   
14581              2.69       5.41                  0.73                5.53   
14582                 0        NaN                  0.83                 NaN   
14583              7.45       0.42                     0                0.24   
14584              8.04        0.6                     0                0.96   

      repitencia_secundaria repitencia_media tama_o_promedio_de_grupo  \
0                     13.76             2.04                      NaN   
1                      8.48             3.16                      NaN   
2                     10.25             1.54                      NaN   
3                     13.84 

In [117]:
# --- LIMPIEZA BASE EDUCACIÓN ---
# Normalizar nombres de columnas
df_raw.columns = df_raw.columns.str.strip().str.lower().str.replace(" ", "_")

# Convertir columnas numéricas (algunas vienen como string desde la API)
columnas_float = ['tasa_matriculación_5_16', 'cobertura_neta']
for col in columnas_float:
    if col in df_raw.columns:
        df_raw[col] = pd.to_numeric(df_raw[col], errors='coerce')

# Convertir año a entero
if 'año' in df_raw.columns:
    df_raw['año'] = pd.to_numeric(df_raw['año'], errors='coerce', downcast='integer')

# 🔁 Corrección Bogotá y San Andrés
if 'departamento' in df_raw.columns:
    df_raw['departamento'] = df_raw['departamento'].str.strip().str.upper()

    # Unificar San Andrés
    df_raw['departamento'] = df_raw['departamento'].str.replace(
        r"ARCHIPI[EÉ]LAGO DE SAN ANDRES[.,\s]*.*PROVIDENCIA Y SANTA CATALINA",
        "ARCHIPIÉLAGO DE SAN ANDRÉS PROVIDENCIA Y SANTA CATALINA",
        regex=True
    )

    # Unificar Bogotá
    df_raw['departamento'] = df_raw['departamento'].str.replace(
        r"\bBOGOT[AÁ][,]?\s*(D\.?C\.?)?\b",
        "BOGOTÁ D.C.",
        regex=True
    )

    # Volver a capitalizar para consistencia
    df_raw['departamento'] = df_raw['departamento'].str.title()

# --- LIMPIEZA BASE DIVIPOLA ---
# Normalizar columnas
df_divipola.columns = df_divipola.columns.str.strip().str.lower().str.replace(" ", "_")

# Renombrar columna conflictiva
if 'tipo:municipio/isla/_área_no_municipalizada' in df_divipola.columns:
    df_divipola = df_divipola.rename(columns={
        'tipo:municipio/isla/_área_no_municipalizada': 'tipo_municipio'
    })

# Corregir coordenadas (longitud y latitud)
if 'longitud' in df_divipola.columns:
    df_divipola['longitud'] = df_divipola['longitud'].astype(str).str.replace(',', '.')
    df_divipola['longitud'] = pd.to_numeric(df_divipola['longitud'], errors='coerce')

if 'latitud' in df_divipola.columns:
    df_divipola['latitud'] = df_divipola['latitud'].astype(str).str.replace(',', '.')
    df_divipola['latitud'] = pd.to_numeric(df_divipola['latitud'], errors='coerce')


In [118]:
df_raw['tasa_matriculaci_n_5_16'] = df_raw['tasa_matriculaci_n_5_16'].astype(float)

In [119]:
import os
import sqlite3


# Crear conexión SQLite (archivo .db en la carpeta Datos)
conn = sqlite3.connect("../Datos/educacion.db")

# Guardar DataFrames como tablas SQL
df_raw.to_sql("educacion_raw", conn, if_exists="replace", index=False)
df_divipola.to_sql("divipola", conn, if_exists="replace", index=False)

print("📁 Tablas 'educacion_raw' y 'divipola' guardadas en SQLite.")


📁 Tablas 'educacion_raw' y 'divipola' guardadas en SQLite.


## Creación de dimensiones

In [120]:
# Dimension de tiempo
# Eliminar si ya existe
conn.execute("DROP TABLE IF EXISTS dim_tiempo;")

# Crear la tabla usando el campo 'a_o'
query_dim_tiempo = """
CREATE TABLE dim_tiempo AS
SELECT 
    ROW_NUMBER() OVER () AS id_tiempo,
    a_o AS anio
FROM educacion_raw
WHERE a_o IS NOT NULL
GROUP BY a_o
ORDER BY a_o;
"""

conn.execute(query_dim_tiempo)

In [131]:
# DIMENSIÓN MUNICIPIO (solo con df_divipola)
dim_municipio = df_divipola[[
    "cod_mpio", "nom_mpio", "cod_dpto", "dpto", "tipo_municipio", "longitud", "latitud"
]].drop_duplicates()

dim_municipio = dim_municipio.rename(columns={
    "cod_mpio": "id_municipio",
    "nom_mpio": "municipio",
    "cod_dpto": "id_departamento",
    "dpto": "departamento"
})

dim_municipio.head(15)

id_municipio              municipio id_departamento departamento  \
0         05001               MEDELLÍN              05    ANTIOQUIA   
1         05002              ABEJORRAL              05    ANTIOQUIA   
2         05004               ABRIAQUÍ              05    ANTIOQUIA   
3         05021             ALEJANDRÍA              05    ANTIOQUIA   
4         05030                  AMAGÁ              05    ANTIOQUIA   
5         05031                 AMALFI              05    ANTIOQUIA   
6         05034                  ANDES              05    ANTIOQUIA   
7         05036            ANGELÓPOLIS              05    ANTIOQUIA   
8         05038              ANGOSTURA              05    ANTIOQUIA   
9         05040                  ANORÍ              05    ANTIOQUIA   
10        05042  SANTA FÉ DE ANTIOQUIA              05    ANTIOQUIA   
11        05044                   ANZÁ              05    ANTIOQUIA   
12        05045               APARTADÓ              05    ANTIOQUIA   
13        05051              ARBOLETES              05    ANTIOQUIA   
14        05055                ARGELIA              05    ANTIOQUIA   

   tipo_municipio   longitud   latitud  
0       Municipio -75.581775  6.246631  
1       Municipio -75.428739  5.789315  
2       Municipio -76.064304  6.632282  
3       Municipio -75.141346  6.376061  
4       Municipio -75.702188  6.038708  
5       Municipio -75.077501  6.909655  
6       Municipio -75.878828  5.657194  
7       Municipio -75.711389  6.109719  
8       Municipio -75.335116  6.885175  
9       Municipio -75.148355  7.074703  
10      Municipio -75.826648  6.556484  
11      Municipio -75.854442  6.302641  
12      Municipio -76.625279  7.882968  
13      Municipio -76.426708  8.849317  
14      Municipio -75.141070  5.731474

In [121]:
# Dimension de municipio
# Eliminar si ya existe
conn.execute("DROP TABLE IF EXISTS dim_municipio;")

# Crear tabla con los nombres correctos desde la API
query_dim_municipio = """
CREATE TABLE dim_municipio AS
SELECT 
    ROW_NUMBER() OVER () AS id_municipio,
    e.c_digo_municipio AS codigo_municipio,
    e.municipio,
    d.dpto AS nombre_departamento,
    d.tipo_municipio,
    CAST(d.longitud AS FLOAT) AS longitud,
    CAST(d.latitud AS FLOAT) AS latitud
FROM educacion_raw e
LEFT JOIN divipola d 
    ON e.c_digo_municipio = d.cod_mpio
WHERE e.c_digo_municipio IS NOT NULL AND e.municipio IS NOT NULL
GROUP BY 
    e.c_digo_municipio, e.municipio, d.dpto, d.tipo_municipio, d.longitud, d.latitud;
"""

conn.execute(query_dim_municipio)


In [122]:
# Dimension de Departamento
# Eliminar si ya existe
conn.execute("DROP TABLE IF EXISTS dim_departamento;")

# Crear la tabla usando nombres corregidos y únicos
query_dim_departamento = """
CREATE TABLE dim_departamento AS
SELECT 
    ROW_NUMBER() OVER () AS id_departamento,
    c_digo_departamento AS codigo_departamento,
    departamento
FROM educacion_raw
WHERE c_digo_departamento IS NOT NULL AND departamento IS NOT NULL
GROUP BY c_digo_departamento, departamento
ORDER BY departamento;
"""

conn.execute(query_dim_departamento)


## Tabla de hechos

In [139]:
# Eliminar si ya existe
conn.execute("DROP TABLE IF EXISTS hechos_educacion_limpia;")

# Crear tabla de hechos con control de duplicados y datos limpios
query_hechos_limpia = """
CREATE TABLE hechos_educacion_limpia AS
SELECT 
    t.id_tiempo,
    m.id_municipio,
    d.id_departamento,
    AVG(CASE 
        WHEN e.tasa_matriculaci_n_5_16 > 100 THEN 100
        WHEN e.tasa_matriculaci_n_5_16 < 0 THEN 0
        ELSE e.tasa_matriculaci_n_5_16
    END) AS tasa_matriculacion,
    AVG(e.cobertura_neta) AS cobertura_neta
FROM educacion_raw e
JOIN dim_tiempo t 
    ON e.a_o = t.anio
JOIN dim_municipio m 
    ON e.c_digo_municipio = m.codigo_municipio AND e.municipio = m.municipio
JOIN dim_departamento d 
    ON e.c_digo_departamento = d.codigo_departamento AND e.departamento = d.departamento
WHERE 
    e.tasa_matriculaci_n_5_16 IS NOT NULL AND
    e.cobertura_neta IS NOT NULL
GROUP BY 
    t.id_tiempo, m.id_municipio, d.id_departamento;
"""

conn.execute(query_hechos_limpia)


# Preguntas

## 1. Respecto a la población del municipio ¿Que porcentaje de escolaridad hay?

In [144]:
query_test = """
SELECT m.municipio, t.anio, h.tasa_matriculacion
FROM hechos_educacion_limpia h
JOIN dim_municipio m ON h.id_municipio = m.id_municipio
JOIN dim_tiempo t ON h.id_tiempo = t.id_tiempo
ORDER BY h.tasa_matriculacion DESC
LIMIT 10;
"""

df = pd.read_sql(query_test, conn)
df

municipio  anio  tasa_matriculacion
0                 Achí  2011               100.0
1              Calamar  2011               100.0
2               Cicuco  2011               100.0
3              Córdoba  2011               100.0
4      Hatillo de Loba  2011               100.0
5               Mompós  2011               100.0
6          San Jacinto  2011               100.0
7  San Juan Nepomuceno  2011               100.0
8                Belén  2011               100.0
9               Caldas  2011               100.0

## 2. ¿Cómo compararía el rendimiento educativo por municipios?

In [141]:
query_2 = """
SELECT 
    m.municipio,
    ROUND(AVG(e.aprobaci_n), 2) AS promedio_aprobacion,
    ROUND(AVG(e.reprobaci_n), 2) AS promedio_reprobacion,
    ROUND(AVG(e.deserci_n), 2) AS promedio_desercion
FROM educacion_raw e
JOIN dim_municipio m ON e.c_digo_municipio = m.codigo_municipio
GROUP BY m.municipio
ORDER BY promedio_aprobacion DESC
LIMIT 10;
"""

df_p2 = pd.read_sql(query_2, conn)
df_p2


municipio  promedio_aprobacion  promedio_reprobacion  \
0     Puerto Arica                98.97                  0.03   
1         Tibirita                98.42                  0.43   
2         La Palma                98.28                  0.20   
3        Guatavita                98.27                  0.27   
4  Miriti - Paraná                98.21                  0.35   
5         Chaguaní                98.19                  0.16   
6         Pupiales                98.10                  0.67   
7        Sutatausa                97.96                  0.11   
8           Aguada                97.72                  0.90   
9            Junín                97.68                  0.98   

   promedio_desercion  
0                1.31  
1                1.16  
2                1.52  
3                1.45  
4                1.70  
5                1.65  
6                1.23  
7                1.93  
8                1.39  
9                1.34

Con base en los datos de rendimiento educativo, se observa que municipios como Puerto Arica, Tibiritá y La Palma presentan los promedios más altos de aprobación, superando el 98%, lo que refleja una gestión académica destacada. Estos municipios también muestran bajos niveles de reprobación (inferiores al 1%) y tasas de deserción entre 1.1% y 1.5%, lo cual sugiere una permanencia escolar relativamente estable. Aunque algunos municipios como Sutatausa y Pupiales mantienen buenos niveles de aprobación, presentan valores levemente superiores en deserción y reprobación comparados con los mejores casos

## 3. ¿Que departamentos son los que mejor cobertura tienen? ¿Pueden hacer cálculo con SQL?

In [134]:
query_cobertura_ponderada_fix = """
SELECT 
    d.departamento,
    ROUND(
        SUM(e.cobertura_neta * e.poblaci_n_5_16) / SUM(e.poblaci_n_5_16), 3
    ) || ' %' AS cobertura_neta_ponderada
FROM educacion_raw e
JOIN dim_departamento d ON e.c_digo_departamento = d.codigo_departamento
WHERE e.cobertura_neta IS NOT NULL AND e.poblaci_n_5_16 > 0
GROUP BY d.departamento
ORDER BY cobertura_neta_ponderada DESC
LIMIT 20;
"""

df_resultado = pd.read_sql_query(query_cobertura_ponderada_fix, conn)
df_resultado


departamento cobertura_neta_ponderada
0             Casanare                 96.974 %
1                Sucre                 95.383 %
2            Santander                 94.022 %
3            Magdalena                 93.709 %
4              Bolívar                 93.511 %
5            Risaralda                 92.527 %
6         Cundinamarca                 92.222 %
7                 Meta                 92.093 %
8                Cesar                 91.978 %
9            Antioquia                 91.238 %
10              Tolima                 91.217 %
11           Atlántico                 91.027 %
12             Córdoba                 90.807 %
13               Huila                 90.007 %
14  Norte De Santander                 89.228 %
15              Boyacá                 89.021 %
16        Bogotá D.C..                 88.626 %
17             Quindio                 88.441 %
18          La Guajira                  84.43 %
19               Cauca                 83.786 %

Los resultados muestran que departamentos como Casanare, Sucre y Santander presentan los niveles más altos de cobertura neta educativa, superando el 94%. Esto indica que una proporción muy alta de la población en edad escolar (5 a 16 años) en estos territorios está accediendo efectivamente al sistema educativo. El análisis tiene en cuenta tanto la cobertura registrada como la población total en ese grupo etario, lo cual permite comparar de forma más justa entre departamentos de distinto tamaño. Esta cobertura ponderada refleja no solo eficiencia en matrícula, sino también una adecuada planificación educativa y atención a la demanda escolar real.